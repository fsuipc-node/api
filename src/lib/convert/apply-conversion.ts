import { VM } from 'vm2';

import { Offset } from '@shared/offset';
import { RawOffsetValue } from '@shared/offset-values';
import { ConvertedOffsetValue } from '@shared/converted-offset-values';

export const applyConversion = (offset: Offset, rawOffsetValue: RawOffsetValue): ConvertedOffsetValue => {
  if (!offset.convert) {
    return rawOffsetValue;
  }

  if (offset.hasMapping) {
    if (offset.isMappingInvalid) {
      return 'INVALID_MAPPING_FUNCTION';
    }

    return offset.mappingFunction(rawOffsetValue);
  } else {
    // tslint:disable-next-line:no-eval
    if (offset.isInvalidConvertExpression) {
      return 'UNSUPPORTED_CONVERSION_EXPRESSION';
    }

    return new VM().run(
      replaceOffsetExpressionValue(offset, rawOffsetValue)
    );
  }
};

export const replaceOffsetExpressionValue = (offset: Offset, rawOffsetValue: RawOffsetValue): string => {
  return offset.convert.replace(
    new RegExp(/{VAL}/g),
    Array.isArray(rawOffsetValue)
      ? `${JSON.stringify(rawOffsetValue)}`
      : typeof rawOffsetValue === 'string'
        ? `'${rawOffsetValue}'`
        : rawOffsetValue.toString()
  );
};
