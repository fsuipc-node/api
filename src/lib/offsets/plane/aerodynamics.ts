import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const APU: OffsetList = {
  incidenceAlpha: new Offset({
    value: 0x2ED0,
    name: 'incidenceAlpha',
    description: 'incidence alpha (body AoA) - radians',
    category: OffsetCategory.AERODYNAMICS,
    type: Type.Double,
    permission: 'r',
  }),
};
