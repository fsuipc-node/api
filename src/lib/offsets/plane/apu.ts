import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const APU: OffsetList = {
  APUGeneratorSwitch: new Offset({
    value: 0xB51,
    name: 'APUGeneratorSwitch',
    category: OffsetCategory.APU,
    description: 'APU generator switch',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  APUGeneratorActive: new Offset({
    value: 0xB52,
    name: 'APUGeneratorActive',
    category: OffsetCategory.APU,
    description: 'APU generator active',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  APUOnFire: new Offset({
    value: 0xB53,
    name: 'APUOnFire',
    category: OffsetCategory.APU,
    description: 'APU generator active',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  APUMaxRPM: new Offset({
    value: 0xB54,
    name: 'APUMaxRPM',
    category: OffsetCategory.APU,
    description: 'APU generator max RPM percent',
    convert: '+({VAL}).toFixed(2)',
    type: Type.Single,
    permission: 'r',
  }),
  APUStarter: new Offset({
    value: 0xB58,
    name: 'APUStarter',
    category: OffsetCategory.APU,
    description: 'APU starter',
    convert: '{VAL} ? 1 : 0',
    type: Type.Single,
    permission: 'rw',
  }),
  APUVoltage: new Offset({
    value: 0xB5C,
    name: 'APUVoltage',
    category: OffsetCategory.APU,
    description: 'APU generator voltage',
    convert: '+({VAL}).toFixed(2)',
    type: Type.Single,
    permission: 'rw',
  }),
};
