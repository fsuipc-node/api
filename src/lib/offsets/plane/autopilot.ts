import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const autopilot: OffsetList = {
  autoPilotAvailable: new Offset({
    value: 0x764,
    name: 'autoPilotAvailable',
    category: OffsetCategory.AUTOPILOT,
    description: 'autopilot available',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  flyByWireELACSwitch: new Offset({
    value: 0x7B6,
    name: 'flyByWireELACSwitch',
    category: OffsetCategory.AUTOPILOT,
    description: 'fly by wire ELAC switch',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  flyByWireELACCompFailFlag: new Offset({
    value: 0x7B7,
    name: 'flyByWireELACCompFailFlag',
    category: OffsetCategory.AUTOPILOT,
    description: 'fly by wire ELAC comp fail flag (read only)',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  flyByWireFACSwitch: new Offset({
    value: 0x7B8,
    name: 'flyByWireFACSwitch',
    category: OffsetCategory.AUTOPILOT,
    description: 'fly by wire FAC switch',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  flyByWireFACCompFailFlag: new Offset({
    value: 0x7B9,
    name: 'flyByWireFACCompFailFlag',
    category: OffsetCategory.AUTOPILOT,
    description: 'fly by wire FAC comp fail flag (read only)',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  flyByWireSECSwitch: new Offset({
    value: 0x7BA,
    name: 'flyByWireSECSwitch',
    category: OffsetCategory.AUTOPILOT,
    description: 'fly by wire SEC switch',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'rw',
  }),
  flyByWireSECCompFailFlag: new Offset({
    value: 0x7BB,
    name: 'flyByWireSECCompFailFlag',
    category: OffsetCategory.AUTOPILOT,
    description: 'fly by wire SEC comp fail flag (read only)',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  apMasterSwitch: new Offset({
    value: 0x7BC,
    name: 'apMasterSwitch',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP master switch',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apWingLevel: new Offset({
    value: 0x7C0,
    name: 'apWingLevel',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP wing level',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apNav1Hold: new Offset({
    value: 0x7C4,
    name: 'apNav1Hold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP NAV1 hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apHeadingHold: new Offset({
    value: 0x7C8,
    name: 'apHeadingHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP heading hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apHeadingValue: new Offset({
    value: 0x7CC,
    name: 'apHeadingValue',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP heading value - degrees',
    convert: 'Math.round(({VAL} * 360) / 65536)',
    type: Type.UInt16,
    permission: 'rw',
  }),
  apAltitudeHold: new Offset({
    value: 0x7D0,
    name: 'apAltitudeHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP altitude hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apAltitudeValue: new Offset({
    value: 0x7D4,
    name: 'apAltitudeValue',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP altitude value - ft',
    convert: '+({VAL} * 3.28084 / 65536).toFixed(2)',
    type: Type.Int32,
    permission: 'rw',
  }),
  apAttitudeHold: new Offset({
    value: 0x7D8,
    name: 'apAttitudeHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP attitude hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apAsHold: new Offset({
    value: 0x7DC,
    name: 'apAsHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP airspeed hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apAsValue: new Offset({
    value: 0x7E2,
    name: 'apAsValue',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP airspeed value - kt',
    type: Type.UInt16,
    permission: 'rw',
  }),
  apMachHold: new Offset({
    value: 0x7E4,
    name: 'apMachHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP mach hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apMachValue: new Offset({
    value: 0x7E8,
    name: 'apMachValue',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP mach value - mach',
    convert: '{VAL} / 65536',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apVsHold: new Offset({
    value: 0x7EC,
    name: 'apVsHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP vertical speed hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apVsValue: new Offset({
    value: 0x7F2,
    name: 'apVsValue',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP vertical speed value - ft/min',
    type: Type.Int16,
    permission: 'rw',
  }),
  apRPMN1Hold: new Offset({
    value: 0x7F4,
    name: 'apRPMN1Hold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP RPM/N1 hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apRPMN1Value: new Offset({
    value: 0x7FA,
    name: 'apRPMN1Value',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP RPM/N1 value - percent',
    convert: 'Math.round({VAL} * 100 / 16384)',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apGlideSlopeHold: new Offset({
    value: 0x7FC,
    name: 'apGlideSlopeHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP GlideSlope hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apApproachHold: new Offset({
    value: 0x800,
    name: 'apApproachHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP Approach hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apBackCourseHold: new Offset({
    value: 0x804,
    name: 'apBackCourseHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP back course hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apYawDamperHold: new Offset({
    value: 0x808,
    name: 'apYawDamperHold',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP yaw damper hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apTOGAAutoThrottle: new Offset({
    value: 0x80C,
    name: 'apTOGAAutoThrottle',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP TO/GA throttle hold',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apAutoThrottleArm: new Offset({
    value: 0x810,
    name: 'apAutoThrottleArm',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP autothrottle arm',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  apMaxBank: new Offset({
    value: 0x2E04,
    name: 'apMaxBank',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP max bank - degrees',
    type: Type.UInt32,
    permission: 'r',
  }),
  apMaxSpeed: new Offset({
    value: 0x2F40,
    name: 'apMaxSpeed',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP max speed',
    type: Type.Double,
    permission: 'r',
  }),
  apCruiseSpeed: new Offset({
    value: 0x2F48,
    name: 'apCruiseSpeed',
    category: OffsetCategory.AUTOPILOT,
    description: 'AP cruise speed',
    type: Type.Double,
    permission: 'r',
  }),
};
