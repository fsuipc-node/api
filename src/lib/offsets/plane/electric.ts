import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const electric: OffsetList = {
  masterBatterySwitch: new Offset({
    value: 0x281C,
    name: 'masterBatterySwitch',
    category: OffsetCategory.ELECTRIC,
    description: 'battery master switch',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'rw',
  }),
  totalLoadAmp: new Offset({
    value: 0x2824,
    name: 'totalLoadAmp',
    category: OffsetCategory.ELECTRIC,
    description: 'total load amps',
    type: Type.Double,
    permission: 'r',
  }),
  batteryLoad: new Offset({
    value: 0x282C,
    name: 'batteryLoad',
    category: OffsetCategory.ELECTRIC,
    description: 'battery load',
    type: Type.Double,
    permission: 'r',
  }),
  batteryVoltage: new Offset({
    value: 0x2834,
    name: 'batteryVoltage',
    category: OffsetCategory.ELECTRIC,
    description: 'battery voltage',
    type: Type.Double,
    permission: 'r',
  }),
  mainBusVoltage: new Offset({
    value: 0x2840,
    name: 'mainBusVoltage',
    category: OffsetCategory.ELECTRIC,
    description: 'main bus voltage',
    type: Type.Double,
    permission: 'r',
  }),
  mainBusAmps: new Offset({
    value: 0x2848,
    name: 'mainBusAmps',
    category: OffsetCategory.ELECTRIC,
    description: 'main bus amps',
    type: Type.Double,
    permission: 'r',
  }),
  avionicsBusVoltage: new Offset({
    value: 0x2850,
    name: 'avionicsBusVoltage',
    category: OffsetCategory.ELECTRIC,
    description: 'avionics bus voltage',
    type: Type.Double,
    permission: 'r',
  }),
  avionicsBusAmps: new Offset({
    value: 0x2858,
    name: 'avionicsBusAmps',
    category: OffsetCategory.ELECTRIC,
    description: 'avionics bus amps',
    type: Type.Double,
    permission: 'r',
  }),
  hotBatteryBusVoltage: new Offset({
    value: 0x2860,
    name: 'hotBatteryBusVoltage',
    category: OffsetCategory.ELECTRIC,
    description: 'hot battery bus voltage',
    type: Type.Double,
    permission: 'r',
  }),
  hotBatteryBusAmps: new Offset({
    value: 0x2868,
    name: 'hotBatteryBusAmps',
    category: OffsetCategory.ELECTRIC,
    description: 'hot battery bus amps',
    type: Type.Double,
    permission: 'r',
  }),
  batteryBusVoltage: new Offset({
    value: 0x2870,
    name: 'batteryBusVoltage',
    category: OffsetCategory.ELECTRIC,
    description: 'battery bus voltage',
    type: Type.Double,
    permission: 'r',
  }),
  batteryBusAmps: new Offset({
    value: 0x2878,
    name: 'batteryBusAmps',
    category: OffsetCategory.ELECTRIC,
    description: 'battery bus amps',
    type: Type.Double,
    permission: 'r',
  }),
  generatorAlternator1BusVoltage: new Offset({
    value: 0x2880,
    name: 'generatorAlternator1BusVoltage',
    category: OffsetCategory.ELECTRIC,
    description: 'generator alternator 1 bus voltage',
    type: Type.Double,
    permission: 'r',
  }),
  generatorAlternator1BusAmps: new Offset({
    value: 0x2888,
    name: 'generatorAlternator1BusAmps',
    category: OffsetCategory.ELECTRIC,
    description: 'generator alternator 1 bus amps',
    type: Type.Double,
    permission: 'r',
  }),
  generatorAlternator2BusVoltage: new Offset({
    value: 0x2890,
    name: 'generatorAlternator2BusVoltage',
    category: OffsetCategory.ELECTRIC,
    description: 'generator alternator 2 bus voltage',
    type: Type.Double,
    permission: 'r',
  }),
  generatorAlternator2BusAmps: new Offset({
    value: 0x2898,
    name: 'generatorAlternator2BusAmps',
    category: OffsetCategory.ELECTRIC,
    description: 'generator alternator 2 bus amps',
    type: Type.Double,
    permission: 'r',
  }),
  generatorAlternator3BusVoltage: new Offset({
    value: 0x28A0,
    name: 'generatorAlternator3BusVoltage',
    category: OffsetCategory.ELECTRIC,
    description: 'generator alternator 3 bus voltage',
    type: Type.Double,
    permission: 'r',
  }),
  generatorAlternator3BusAmps: new Offset({
    value: 0x28A8,
    name: 'generatorAlternator3BusAmps',
    category: OffsetCategory.ELECTRIC,
    description: 'generator alternator 3 bus amps',
    type: Type.Double,
    permission: 'r',
  }),
  generatorAlternator4BusVoltage: new Offset({
    value: 0x28B0,
    name: 'generatorAlternator4BusVoltage',
    category: OffsetCategory.ELECTRIC,
    description: 'generator alternator 4 bus voltage',
    type: Type.Double,
    permission: 'r',
  }),
  generatorAlternator4BusAmps: new Offset({
    value: 0x28B8,
    name: 'generatorAlternator4BusAmps',
    category: OffsetCategory.ELECTRIC,
    description: 'generator alternator 4 bus amps',
    type: Type.Double,
    permission: 'r',
  }),
};
