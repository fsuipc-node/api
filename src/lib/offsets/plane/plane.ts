import { Type } from 'fsuipc';

import { OffsetCategory } from '@shared/offset-category';
import { OffsetList } from '@shared/offset-list';
import { Offset } from '@shared/offset';

export const plane: OffsetList = {
  vs0: new Offset({
    value: 0x538,
    name: 'vs0',
    category: OffsetCategory.PLANE,
    description: 'Stall speed full flaps - kt',
    convert: 'ftsecToKt',
    mapping: true,
    type: Type.Double,
    permission: 'r',
  }),
  vs1: new Offset({
    value: 0x540,
    name: 'vs1',
    category: OffsetCategory.PLANE,
    description: 'Stall speed clean - kt',
    convert: 'ftsecToKt',
    mapping: true,
    type: Type.Double,
    permission: 'r',
  }),
  vc: new Offset({
    value: 0x548,
    name: 'vc',
    category: OffsetCategory.PLANE,
    description: 'Cruise speed - kt',
    convert: 'ftsecToKt',
    mapping: true,
    type: Type.Double,
    permission: 'r',
  }),
  vmd: new Offset({
    value: 0x550,
    name: 'vmd',
    category: OffsetCategory.PLANE,
    description: 'Minimum drag speed - kt',
    convert: 'ftsecToKt',
    mapping: true,
    type: Type.Double,
    permission: 'r',
  }),
  smokeSystemControl: new Offset({
    value: 0x5D8,
    name: 'smokeSystemControl',
    category: OffsetCategory.PLANE,
    convert: '!!{VAL}',
    description: 'smoke system control on/off',
    type: Type.UInt16,
    permission: 'rw',
  }),
  retractableGear: new Offset({
    value: 0x60C,
    name: 'retractableGear',
    category: OffsetCategory.PLANE,
    description: 'retractable gear (read only)',
    convert: '!!{VAL}',
    type: Type.Byte,
    permission: 'r',
  }),
  retractableLeftFloatExtension: new Offset({
    value: 0x614,
    name: 'retractableLeftFloatExtension',
    category: OffsetCategory.PLANE,
    description: 'retractable left float extension - percent',
    convert: 'Math.floor({VAL} / 16384 * 100)',
    type: Type.UInt16,
    permission: 'r',
  }),
  retractableRightFloatExtension: new Offset({
    value: 0x616,
    name: 'retractableRightFloatExtension',
    category: OffsetCategory.PLANE,
    description: 'retractable right float extension - percent',
    convert: 'Math.floor({VAL} / 16384 * 100)',
    type: Type.UInt16,
    permission: 'r',
  }),
  hasFlaps: new Offset({
    value: 0x778,
    name: 'hasFlaps',
    category: OffsetCategory.PLANE,
    description: 'has flaps',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasStallHorn: new Offset({
    value: 0x77C,
    name: 'hasStallHorn',
    category: OffsetCategory.PLANE,
    description: 'has stall horn',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasMixtureEngine: new Offset({
    value: 0x780,
    name: 'hasMixtureEngine',
    category: OffsetCategory.PLANE,
    description: 'has mixture engine',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasCarbHeat: new Offset({
    value: 0x784,
    name: 'hasCarbHeat',
    category: OffsetCategory.PLANE,
    description: 'has carburator heater',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasSpoilers: new Offset({
    value: 0x78C,
    name: 'hasSpoilers',
    category: OffsetCategory.PLANE,
    description: 'has spoilers',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  isTailDragger: new Offset({
    value: 0x790,
    name: 'isTailDragger',
    category: OffsetCategory.PLANE,
    description: 'has spoilers',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasStrobes: new Offset({
    value: 0x794,
    name: 'hasStrobes',
    category: OffsetCategory.PLANE,
    description: 'has strobes',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  hasToeBrakes: new Offset({
    value: 0x79C,
    name: 'hasToeBrakes',
    category: OffsetCategory.PLANE,
    description: 'has toe brakes',
    convert: '!!{VAL}',
    type: Type.UInt32,
    permission: 'r',
  }),
  maxMach: new Offset({
    value: 0xB0C,
    name: 'maxMach',
    category: OffsetCategory.PLANE,
    description: 'mach maximum operating - not FS2004',
    convert: '+({VAL} / 20438).toFixed(2)',
    type: Type.Int32,
    permission: 'r',
  }),
  fuelSelectorsCount: new Offset({
    value: 0x123E,
    name: 'fuelSelectorsCount',
    category: OffsetCategory.PLANE,
    description: 'number of fuel selectors for the aircraft',
    type: Type.Byte,
    permission: 'r',
  }),
  aircraftEmptyWeight: new Offset({
    value: 0x1330,
    name: 'aircraftEmptyWeight',
    category: OffsetCategory.PLANE,
    description: 'aicraft empty weight',
    convert: '+({VAL} / 256).toFixed(2)',
    type: Type.Int32,
    permission: 'r',
  }),
  aircraftMaxGrossWeight: new Offset({
    value: 0x1334,
    name: 'aircraftMaxGrossWeight',
    category: OffsetCategory.PLANE,
    description: 'aicraft max gross weight',
    convert: '+({VAL} / 256).toFixed(2)',
    type: Type.Int32,
    permission: 'r',
  }),
};
