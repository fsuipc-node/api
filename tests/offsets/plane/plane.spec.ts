import { plane as offsets } from '@offsets/plane/plane';
import { replaceOffsetExpressionValue } from '@convert/apply-conversion';

describe('offset - plane/plane', () => {
  const offsetsTestCases = [
    { name: 'smokeSystemControl', value: 0, expectedResult: false },
    { name: 'smokeSystemControl', value: 1, expectedResult: true },
    { name: 'retractableGear', value: 0, expectedResult: false },
    { name: 'retractableGear', value: 1, expectedResult: true },
    { name: 'retractableLeftFloatExtension', value: 13200, expectedResult: 80 },
    { name: 'retractableRightFloatExtension', value: 13200, expectedResult: 80 },
    { name: 'hasFlaps', value: 0, expectedResult: false },
    { name: 'hasFlaps', value: 1, expectedResult: true },
    { name: 'hasStallHorn', value: 0, expectedResult: false },
    { name: 'hasStallHorn', value: 1, expectedResult: true },
    { name: 'hasSpoilers', value: 0, expectedResult: false },
    { name: 'hasSpoilers', value: 1, expectedResult: true },
    { name: 'isTailDragger', value: 0, expectedResult: false },
    { name: 'isTailDragger', value: 1, expectedResult: true },
    { name: 'hasStrobes', value: 0, expectedResult: false },
    { name: 'hasStrobes', value: 1, expectedResult: true },
    { name: 'hasToeBrakes', value: 0, expectedResult: false },
    { name: 'hasToeBrakes', value: 1, expectedResult: true },
    { name: 'maxMach', value: 13200, expectedResult: 0.65 },
    { name: 'aircraftEmptyWeight', value: 512060, expectedResult: 2000.23 },
    { name: 'aircraftMaxGrossWeight', value: 512060, expectedResult: 2000.23 },
  ];

  describe('offsets list', () => {
    it('should have required properties', () => {
      expect(offsets).toMatchSnapshot();
    });
  });

  offsetsTestCases.forEach(testedOffset => {
    describe(testedOffset.name, () => {
      it('should convert data properly', () => {
        const convertExpression = replaceOffsetExpressionValue(offsets[testedOffset.name], testedOffset.value);

        // tslint:disable-next-line:no-eval
        expect(eval(convertExpression)).toEqual(testedOffset.expectedResult);
      });
    });
  });
});
