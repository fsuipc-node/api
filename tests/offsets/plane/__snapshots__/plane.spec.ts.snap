// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`offset - plane/plane offsets list should have required properties 1`] = `
Object {
  "aircraftEmptyWeight": Offset {
    "category": "plane",
    "convert": "+({VAL} / 256).toFixed(2)",
    "description": "aicraft empty weight",
    "length": undefined,
    "mapping": undefined,
    "name": "aircraftEmptyWeight",
    "permission": "r",
    "type": 3,
    "value": 4912,
  },
  "aircraftMaxGrossWeight": Offset {
    "category": "plane",
    "convert": "+({VAL} / 256).toFixed(2)",
    "description": "aicraft max gross weight",
    "length": undefined,
    "mapping": undefined,
    "name": "aircraftMaxGrossWeight",
    "permission": "r",
    "type": 3,
    "value": 4916,
  },
  "fuelSelectorsCount": Offset {
    "category": "plane",
    "convert": undefined,
    "description": "number of fuel selectors for the aircraft",
    "length": undefined,
    "mapping": undefined,
    "name": "fuelSelectorsCount",
    "permission": "r",
    "type": 0,
    "value": 4670,
  },
  "hasCarbHeat": Offset {
    "category": "plane",
    "convert": "!!{VAL}",
    "description": "has carburator heater",
    "length": undefined,
    "mapping": undefined,
    "name": "hasCarbHeat",
    "permission": "r",
    "type": 6,
    "value": 1924,
  },
  "hasFlaps": Offset {
    "category": "plane",
    "convert": "!!{VAL}",
    "description": "has flaps",
    "length": undefined,
    "mapping": undefined,
    "name": "hasFlaps",
    "permission": "r",
    "type": 6,
    "value": 1912,
  },
  "hasMixtureEngine": Offset {
    "category": "plane",
    "convert": "!!{VAL}",
    "description": "has mixture engine",
    "length": undefined,
    "mapping": undefined,
    "name": "hasMixtureEngine",
    "permission": "r",
    "type": 6,
    "value": 1920,
  },
  "hasSpoilers": Offset {
    "category": "plane",
    "convert": "!!{VAL}",
    "description": "has spoilers",
    "length": undefined,
    "mapping": undefined,
    "name": "hasSpoilers",
    "permission": "r",
    "type": 6,
    "value": 1932,
  },
  "hasStallHorn": Offset {
    "category": "plane",
    "convert": "!!{VAL}",
    "description": "has stall horn",
    "length": undefined,
    "mapping": undefined,
    "name": "hasStallHorn",
    "permission": "r",
    "type": 6,
    "value": 1916,
  },
  "hasStrobes": Offset {
    "category": "plane",
    "convert": "!!{VAL}",
    "description": "has strobes",
    "length": undefined,
    "mapping": undefined,
    "name": "hasStrobes",
    "permission": "r",
    "type": 6,
    "value": 1940,
  },
  "hasToeBrakes": Offset {
    "category": "plane",
    "convert": "!!{VAL}",
    "description": "has toe brakes",
    "length": undefined,
    "mapping": undefined,
    "name": "hasToeBrakes",
    "permission": "r",
    "type": 6,
    "value": 1948,
  },
  "isTailDragger": Offset {
    "category": "plane",
    "convert": "!!{VAL}",
    "description": "has spoilers",
    "length": undefined,
    "mapping": undefined,
    "name": "isTailDragger",
    "permission": "r",
    "type": 6,
    "value": 1936,
  },
  "maxMach": Offset {
    "category": "plane",
    "convert": "+({VAL} / 20438).toFixed(2)",
    "description": "mach maximum operating - not FS2004",
    "length": undefined,
    "mapping": undefined,
    "name": "maxMach",
    "permission": "r",
    "type": 3,
    "value": 2828,
  },
  "retractableGear": Offset {
    "category": "plane",
    "convert": "!!{VAL}",
    "description": "retractable gear (read only)",
    "length": undefined,
    "mapping": undefined,
    "name": "retractableGear",
    "permission": "r",
    "type": 0,
    "value": 1548,
  },
  "smokeSystemControl": Offset {
    "category": "plane",
    "convert": "!!{VAL}",
    "description": "smoke system control on/off",
    "length": undefined,
    "mapping": undefined,
    "name": "smokeSystemControl",
    "permission": "rw",
    "type": 5,
    "value": 1496,
  },
  "vc": Offset {
    "category": "plane",
    "convert": "ftsecToKt",
    "description": "Cruise speed - kt",
    "length": undefined,
    "mapping": true,
    "name": "vc",
    "permission": "r",
    "type": 8,
    "value": 1352,
  },
  "vmd": Offset {
    "category": "plane",
    "convert": "ftsecToKt",
    "description": "Minimum drag speed - kt",
    "length": undefined,
    "mapping": true,
    "name": "vmd",
    "permission": "r",
    "type": 8,
    "value": 1360,
  },
  "vs0": Offset {
    "category": "plane",
    "convert": "ftsecToKt",
    "description": "Stall speed full flaps - kt",
    "length": undefined,
    "mapping": true,
    "name": "vs0",
    "permission": "r",
    "type": 8,
    "value": 1336,
  },
  "vs1": Offset {
    "category": "plane",
    "convert": "ftsecToKt",
    "description": "Stall speed clean - kt",
    "length": undefined,
    "mapping": true,
    "name": "vs1",
    "permission": "r",
    "type": 8,
    "value": 1344,
  },
}
`;
